Document behaviour of ternary operator of literals. (#12708)
ethereum/solidity: docs/types/operators.rst

   operand, use the type of the right operand,
3. otherwise, the operation is not allowed.

> .. Warning::

Reviewer: The [section about literals](https://docs.soliditylang.org/en/latest/types.html#rational-and-integer-literals) needs clarification too.

First, it says `by using them together with a non-literal expression` and to me this does not imply that `true` and `false` count too. They are not "non-literal expressions".

Second, I think this section needs a warning too, pointing out the behavior of the ternary operator. I'd mention the case of array literals too (i.e. something like `[255][0]` being of `uint8` type and not a literal expression).

> .. Warning::

Reviewer: Aren't these case-sensitive?
```suggestion
.. warning::
```

    The "type of the operand" is calculated through a function mobileType() that assigns, in case we're dealing with Rational Literals,
    the largest possible type that is smaller than the size of the value. An unintended effect of this mechanism is that rational literals
    ( that are not floats ) get converted to the largest possible integer type that just covers the size of the number.

    For eg, a Rational literal 100 will get converted to uint8.

>     Changes like these can lead to overflow errors and other unintended consequences down the line during various operations.

Reviewer: The C++ `mobileType()` function is an implementation detail so I don't think it should be mentioned in the docs.

Also, we do not have floats :) We do have fixed-point types but something like `1.23` really is just a rational literal and in our type system has the same type as integer literals. Such literals have unlimited precision so it's neither fixed- nor float-point. It's an exact representation.

I'd put that warning lower, either just below `If both are literal numbers, the operation is computed with arbitrary precision.` or at the end of the section. That paragraph already explains what a mobile type is.

And here's how I would reword it:

```rst
.. warning::
    The result of the ternary operator is never a literal, even if all of its operands are literals.
    The result type is the bigger of the mobile types of the integer operands.

    For example ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.
    The reason is that ``(true ? 1 : 0)`` is of ``uint8`` type, which forces the addition to be performed in ``uint8`` as well,
    and 256 exceeds the range allowed for this type.

    Another consequence is that an expression like ``1.5 + 1.5`` is valid but ``1.5 + (true ? 1.5 : 2.5)`` is not.
    This is because the former is a literal expression evaluated in unlimited precision and only its final value matters.
    The latter involves a conversion of a fractional rational literal to an integer, which is currently disallowed.
```

The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
left operand for the operation and the result.

> .. note::

Reviewer 2: I'm sorry - I just now opened this page in a wider context, and I think this should just be a section
```
Ternary Operator
--------------
```
instead of a note.

> .. warning::

Reviewer 2: Some suggestions:
 - please check how we call the operator in other parts of the documentation and at least also add `? ... : ...` as an alternative here, so that everyone knows what we are talking about
 - "literal" is more strict than what we want here. I think we use the term rational constant or something, please check.

At this point in the documentation, I think a "note" is better then a warning - we don't use a warning for `**` either.

>     The result of the ternary operator ``... ? ... : ...`` is never a rational, even if all of its operands are rationals.

Reviewer: To be honest, the terminology around literals/constants is a bit unclear to me too. I'd say at least "rational constant" and I saw @chriseth suggest that. Do we actually refer to them as just "rationals"?
```suggestion
    The result of the ternary operator (``... ? ... : ...``) is never a rational constant, even if all of its operands are rational constants.
```

Reviewer: Actually, after looking at the other file, maybe we should go with "number literal"/"number literal expression" for consistency?

Author: I'm pushing changes with "number literal" .
We can replace it further if we decide so

>     The result type is the bigger of the mobile types of the integer operands.

Reviewer: I originally suggested "integer", but maybe "numeric" would be more fitting here? It applies to fractional numbers too.
```suggestion
    The result type is the bigger of the mobile types of the numeric operands.
```

>     The result type is the bigger of the mobile types of the numeric operands.

Reviewer 2: ```suggestion
    The result type is determined from the types of the two operands in the same way as above, converting to their mobile type first if required.
```

Reviewer 2: The main reason to rephrase this is that "size" is difficult because of signed types.

>     For example ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.

Reviewer 2: ```suggestion
    As a consequence, ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.
```

The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
left operand for the operation and the result.

Ternary Operator
----------------

> The result of the ternary operator (``... ? ... : ...``) is never a number literal, even if all of its operands are number literals.

Reviewer 2: Now that this is its own section, maybe we should start with a more generic introduction unrelated to number literals.

>     The result type is determined from the types of the two operands in the same way as above, converting to their mobile type first if required.

Reviewer: This will probably not render properly if you indent the text like this. The header is enough to create a section, you don't have to indent the content in RST or Markdown. The indented text may be interpreted as a code block.

The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
left operand for the operation and the result.

Ternary Operator
----------------
The ternary operator is used in expressions of the form ``<expression> ? <trueExpression> : <falseExpression>``.

> It executes one of the two given expressions depending upon the result of the evaluation of the main ``expression``.
> If ``expression`` evaluates to True, then ``<trueExpression>`` will be executed, else, ``<falseExpression>`` is executed.

Reviewer 3: ```suggestion
It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``expression``.
If ``expression`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
```

> It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``expression``.
> If ``expression`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.

Reviewer 3: ```suggestion
It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``<expression>``.
If ``<expression>`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
```
Otherwise it's inconsistend with the use of ``<trueExpression>`` and ``<falseExpression>`` isn't it?

> The result of the ternary operator cannot have a rational number type, even if all of its operands are rational number literals.

Reviewer 3: ```suggestion
The result of the ternary operator does not have a rational number type, even if all of its operands are rational number literals.
```
:-)?

> The result of the ternary operator (``... ? ... : ...``) is never a number literal, even if all of its operands are number literals.

Reviewer 3: ```suggestion
The result of the ternary operator is never a number literal, even if all of its operands are number literals.
```
We don't need to remention the form of ternary expressions now that we started off with it.

But I'm also not sure if we should say that the result "is never a number literal"... That's a bit of a weird thing to say: a complex expression obviously isn't a literal... technically, it would be accurate to say "never has rational number type", but I'm not sure that really makes things clear, so I'm not sure what's best to do here...

The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
left operand for the operation and the result.

Ternary Operator
----------------
The ternary operator evaluates a given expression to a true or false value and then

> executes one of the two given statements depending upon the result of the evaluation.

Reviewer 3: Technically, what you call ``statementA`` and ``statementB`` aren't statements, but also expressions...
